torch/csrc/autograd/engine.cpp

```diff
@@ -112,134 +122,132 @@ auto Engine::thread_on_exception(FunctionTask& task, std::exception& e) -> void
   }
 }
 
-static variable_list call_pre_hooks(Function& fn, variable_list grad_output) {
+static variable_list call_pre_hooks(Function& fn, variable_list inputs) {
   for (auto& hook : fn.pre_hooks) {
-    grad_output = (*hook)(grad_output);
+    inputs = (*hook)(inputs);
   }
-  return grad_output;
+  return inputs;
 }
 
-static variable_list call_post_hooks(Function& fn, variable_list grad_input, variable_list grad_output) {
+static variable_list call_post_hooks(Function& fn, variable_list outputs, variable_list inputs) {
   for (auto& hook : fn.post_hooks) {
-    grad_input = (*hook)(grad_input, grad_output);
+    outputs = (*hook)(outputs, inputs);
   }
-  return grad_input;
+  return outputs;
 }
 
-static variable_list call_function(FunctionTask& task) {
-  auto grad_output = call_pre_hooks(*task.fn, GradBuffer::variables(std::move(task.grad)));
-  auto grad_input = task.fn->apply(grad_output);
-  return call_post_hooks(*task.fn, std::move(grad_input), std::move(grad_output));
+static std::pair<bool, variable_list> call_function(FunctionTask& task) {
+  auto& fn = *task.fn;
+  auto inputs = call_pre_hooks(fn, InputBuffer::variables(std::move(task.inputs)));
+
+  auto& function_callbacks = task.base->function_callbacks;
```

Comment:
Ok, that makes sense, I think this explanation should be clearly stated in the doc.
Post hooks are not documented right?